odb_extractor_atoms.py

```diff
@@ -59,98 +223,297 @@ def select_obsclass(classes):
+    """Analyze a json observing sequence from the ODB and define atoms."""
 
     classes = []
+    guiding = []
     qastates = []
     atoms = []
     natom = 0
-    nabba = 0
 
     # Make dictionary out of obslog to get QA state
     obslog = {}
-    for log_entry in observation['obsLog']:
-        obslog[log_entry['label']] = {'qaState': log_entry['qaState'], 'filename': log_entry['filename']}
-    datalabels = list(obslog.keys())
+    datalabels = []
+    if 'obsLog' in observation.keys():
+        for log_entry in observation['obsLog']:
+            obslog[log_entry['label']] = {'qaState': log_entry['qaState'], 'filename': log_entry['filename']}
+        datalabels = list(obslog.keys())
 
     # Sequence analysis
     sequence = observation['sequence']
     nsteps = len(sequence)
 
+    # First pass to analyze offsets and exptimes/coadds
+    do_not_split = False
+    exptimes = []
+    coadds = []
+    qoffsets = []
+    poffsets = []
+    qoffsets_sky = []
+    poffsets_sky = []
+    config = {'inst': '', 'fpu': [], 'disperser': [], 'filter': [], 'wavelength': []}
+    for ii, step in enumerate(sequence):
+        step_keys = list(step.keys())
+
+        inst = step['instrument:instrument']
+        #     print(inst, fpuinst[inst])
+        if inst == 'Visitor Instrument':
+            inst = step['instrument:name'].split(' ')[0]
+            if inst in ["'Alopeke", "Zorro"]:
+                fpu = 'None'
+            else:
+                fpu = inst
+        else:
+            fpu = step[fpuinst[inst]]
+        config['inst'] = inst
+        config['fpu'].append(fpu)
+
+        if 'instrument:disperser' in step_keys:
+            disperser = step['instrument:disperser']
+        elif inst in ['IGRINS', 'MAROON-X']:
+            disperser = inst
+        else:
+            disperser = 'None'
+        if inst == 'GNIRS':
+            if step['instrument:acquisitionMirror'] == 'in' and \
+                    step['instrument:decker'] == 'acquisition':
+                disperser = 'mirror'
+            else:
+                disperser = disperser.strip('grating') + step['instrument:crossDispersed']
+        elif inst == 'Flamingos2' and fpu == 'FPU_NONE':
+            if step['instrument:decker'] == 'IMAGING':
+                disperser = step['instrument:decker']
+        config['disperser'].append(disperser)
+
+        if 'instrument:filter' in step_keys:
+            filter = step['instrument:filter']
+        elif inst == 'GPI':
+            filter = find_filter(fpu, gpi_filter_wav)
+        else:
+            if inst == 'GNIRS':
+                filter = 'None'
+            else:
+                filter = 'Unknown'
+        if inst == 'NIFS' and 'Same as Disperser' in filter:
+            for filt in list(nifs_filter_wav.keys()):
+                if disperser[0] in filt:
+                    filter = filt
+                    break
+        config['filter'].append(filter)
```

Comment:
In general, this function is making me a bit nervous. I'm sure we can assume that all of these fields will exist in the sequence and be populated by the ODB, but I feel like the whole thing should be wrapped in a `try`...`except` block:
```
try:
   ...
except KeyError as e:
   ...
```
As it currently stands, if for some obscure reason, one of them doesn't exist, we are either going to get a corrupt data structure that will have unpredictable side effects, or we will get an uncaught error that is going to bubble up and crash the entire Scheduler.